### ISSUE TITLE
Unable to stopAndUnloadAsync a recording that just got started

### ISSUE THREAD

**Issue by James2516** [expo/expo#1709]
### Reproducible Demo
https://snack.expo.io/rJ6j67G6f

After calling `await recording.startAsync()`, it needs at least ~300ms before `recording.stopAndUnloadAsync()` can be called, otherwise it'll fail and subsequent `startAsync` calls will result in 
> error: Only one Recording object can be prepared at a given time.

Tested on Android.

**Comment by Kumagor0** [comment]
I am having the same problem with Expo 32.0.2 on Android.

If I call recording.stopAndUnloadAsync() less than 120ms after recording.startAsync() is resolved, I get the following error:

Stop encountered an error: recording not stopped
- node_modules/react-native/Libraries/BatchedBridge/NativeModules.js:95:55 in <unknown>
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:397:4 in __invokeCallback
- ... 4 more stack frames from framework internals

**Comment by pexea12** [comment]
I am facing the same issue no matter how long I wait.

**Comment by nhim175** [comment]
I'm facing this issue too. It says: `Cannot unload a Recording that has not been prepared.`

**Comment by kashsbd** [comment]
hello guys, I face the same error. Did you solve the error ?

**Comment by cwilby** [comment]
Similar issue - `Cannot unload a Recording that has not been prepared.`

**Comment by cwilby** [comment]
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L349-L362

The error is thrown when `this._canRecord` is `false`.

`this._canRecord` is set to false:

- When it's declared: https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L173

- When ` _cleanupForUnloadedRecorder` is called:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L185

`this._canRecord` is set to true:

- When `prepareToRecordAsync` is called to prepare the recording:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L330

`this._canRecord` is referenced:

- Before polling:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L220

- During a poll:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L199

- When performing any recording operation:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L232-L243

- When getting the recording status:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L249-L262

- When preparing to record:
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L310

- When stopping and unloading a recording (which this issue is discussing):
https://github.com/expo/expo/blob/f6d67bf6ede4a4ea5ec36d376e669aa6021d0f56/packages/expo-av/src/Audio/Recording.ts#L350

**Comment by cwilby** [comment]
My issue was code-related.

```
const { recording: { stopAndUnloadAsync } } = this;

await stopAndUnloadAsync(); // causes error
```

vs

```
await this.recording.stopAndUnloadAsync();
```

Don't destructure functions!

**Comment by zlwu** [comment]
I met the same problem when using Audio.RECORDING_OPTIONS_PRESET_HIGH_QUALITY config. But when I switch to prepareToRecordAsync(Audio.RECORDING_OPTIONS_PRESET_HIGH_QUALITY), this issue got fixed.

**Comment by galzate-blumer** [comment]
Some updates about this issue? I'm having this issue on latest version and I haven't way to stop a recording

**Comment by kjanjuha** [comment]
I'm still having this issue too. Any solutions? It is something to do with the fact that its all asynchronous

**Comment by tanbt** [comment]
I also got this problem so far, `stopAndUnloadAsync` just keep waiting, no additional message.

**Comment by tanbt** [comment]
I made it work by seriously consider the [docs](https://docs.expo.io/versions/latest/sdk/audio/)

> Note that only one recorder is allowed to exist in the state between prepareToRecordAsync and stopAndUnloadAsync at any given time.

by introducing a hook for recording instance `const [recording, setRecording] = useState<Audio.Recording>();`.

Here's the full source: https://gist.github.com/tanbt/857da772db304c0f1562b28c823a0777 (just for the recording feature).

**Comment by tanbt** [comment]
@IjzerenHein So I can say this isn't a bug, but the docs' example can be improved
https://docs.expo.io/versions/latest/sdk/audio/#example-2

**Comment by IjzerenHein** [comment]
Hi! I've taken a closer look at this and updated the snack to work with SDK 38 and add some additional error handling.

https://snack.expo.io/@ijzerenhein/43e79d

What appears to be going on, is that when `stopAndUnloadAsync` is called too quickly after start, it fails with the error `recording not stopped`. In the snack, if you were to hit "Record" again after that it would throw the `Only one Recording object can be prepared at a given time.` error because the previous recording has not been cleaned up correctly.

After digging around, it seems that the MediaRecorder class on Android will purposely cause `stop` to fail when the recording has not yet received any valid recording data. This article gives a good explanation: https://www.hiren.dev/2017/01/android-media-recorder-stop-failed-1007.html

### 1) You called stop method too early.

According to documentation on MediaRecorder.java class it says

Stops recording. Call this after start(). Once recording is stopped, you will have to configure it again as if it has just been constructed. Note that a RuntimeException is intentionally thrown to the application, if no valid audio/video data has been received when stop() is called. This happens if stop() is called immediately after start(). The failure lets the application take action accordingly to clean up the output file (delete the output file, for instance), since the output file is not properly constructed when this happens.

That means if you start recording and in couple of seconds you call stop recording there is no significant data of recording hence stop method throws an exception. So you must wait for sometime. From my testing I found out that you should at least for minimum 10 seconds. So the solution is to have stop method called inside try catch block and handle the exception properly or do not allow user to stop recording till 10 seconds of start of recording. You can either disable stop button and make it enable after sometime. Once the exception is thrown, you should clean the the resources, release media recorder and camera and all other objects.


So in short, this behaviour is sort of by design in Android. But unfortunately expo-av is not handling this case or giving an appropriate error. We'll have a look on how to address this.

**Comment by Aryk** [comment]
For anyone who needs this fix on managed until the SDK 40 release comes out...

```ts
          try {
            await recordRef.current.stopAndUnloadAsync();
          } catch (e) {
            if (e.message.includes("Stop encountered an error: recording not stopped")) {
              await ExponentAV.unloadAudioRecorder();
              await recordRef.current._cleanupForUnloadedRecorder({durationMillis: 0} as any);
            } else {
              await handleError(e, {userMessage: "An error occurred stopping the recording."});
            }
          } finally {
            recordRef.current = undefined;
            timerStop();
            success = true;
          }
```

**Comment by kidfrom** [comment]
@Aryk could you tell me more about `ExponentAV.unloadAudioRecorder();`?


I use `await recordingInstance._cleanupForUnloadedRecorder();` and it solves the problem.
```
  /**
   * // TODO: fix Error: Cannot unload a Recording that has already been unloaded. || Error: Cannot unload a Recording that has not been prepared.
   * Steps to reproduce: Spam the 'Press to record the audio' button
   * Severity: Minor
   *
   * https://github.com/expo/expo/issues/1709
   */
  const recorderStop = async () => {
    await new Promise((resolve) => setTimeout(resolve, 1000)); // handle Error: Stop encountered an error: recording not stopped
    const status = await recorder.getStatusAsync();
    if (status.isRecording === true) {
      try {
        await recorder.stopAndUnloadAsync();
        console.log("recorder stopped");
      } catch (error) {
        if (
          error.message.includes(
            "Stop encountered an error: recording not stopped"
          )
        ) {
          await recorder._cleanupForUnloadedRecorder({
            canRecord: false,
            durationMillis: 0,
            isRecording: false,
            isDoneRecording: false,
          });
          console.log(`recorderStop() error handler: ${error}`);
        } else if (
          error.message.includes(
            "Cannot unload a Recording that has already been unloaded."
          ) ||
          error.message.includes(
            "Error: Cannot unload a Recording that has not been prepared."
          )
        ) {
          console.log(`recorderStop() error handler: ${error}`);
        } else {
          console.error(`recorderStop(): ${error}`);
        }
      }
    }
  };
```

**Comment by Aryk** [comment]
> @Aryk could you tell me more about `ExponentAV.unloadAudioRecorder();`?
> 
> I use `await recordingInstance._cleanupForUnloadedRecorder();` and it solves the problem.
 

Look at the stopAndUnloadAsync function. I added it because I saw it there as well.
